Pass a constructor to assert_throws_dom and promise_rejects_dom.

bzbarsky opened the pull request:
Then we can verify that the DOMExceptions come from the right global.

Fixes https://github.com/web-platform-tests/wpt/issues/21809

bzbarsky committed:
Pass a constructor to assert_throws_dom and promise_rejects_dom.

Then we can verify that the DOMExceptions come from the right global.

Fixes https://github.com/web-platform-tests/wpt/issues/21809

bzbarsky commented:
Note: I have run this through Firefox CI, but I would be pretty surprised if there were no failures in Chrome or Safari.  I just don't know how to run the CI bits for those without creating a PR first....

annevk commented on resources/testharness.js:
```diff
@@ -1582,6 +1583,12 @@ policies and contribution forms [3].
                    {type:type});
 
+            // Check that the exception is from the right global
+            assert(e.constructor == constructor,
```
```suggestion
            assert(e.constructor === constructor,
```
No?

bzbarsky replied:
Good catch, done.

bzbarsky replied:
Good catch, done.

annevk reviewed (approved):
This is wonderful!

cc @domenic

stephenmcgruer commented:
Single new failure from Chrome: https://wpt.fyi/results/css/cssom/CSSStyleSheet-constructable.html?diff&filter=ADC&run_id=421440008&run_id=407660013

6 new failures from Safari: https://wpt.fyi/results/fetch/api/abort/serviceworker-intercepted.https.html?diff&filter=ADC&run_id=440070009&run_id=430610010

Any or all could be real bugs uncovered, of course :)

bzbarsky commented:
OK, the new Safari failures on `fetch/api/abort/serviceworker-intercepted.https.html` are due to the Safari bug that came up a few times already: `fetch` in Safari throws exceptions from the caller global in some way, not from the global of the `fetch` method or the `fetch` method's `this` object.

The new Chrome failure on `css/cssom/CSSStyleSheet-constructable.html` is due to a missing ctor argument; Firefox CI did not catch that because we don't have enough of constructable stylesheets implemented yet.

bzbarsky commented:
And of course it's possible that there are cases where all browsers currently do the wrong thing and the test now tests for that wrong thing... but short of carefully and manually editing all callsites, I don't know how to deal with that.

The fact that the global is optional, so it's easy to assume `self` is not great, but for the `assert_throws_js` case people are likely to just type `TypeError` without thinking overmuch about it, so I don't think it's any worse than that in practice.

bzbarsky commented:
Working on the API test failure.

bzbarsky commented:
OK, I fixed the API tests and added some that specifically test the `assert_throws_dom` global bits.  I also moved the global check to the end of `assert_throws_dom_impl` so that more-specific errors are reported first if the exception thrown is not a `DOMException` at all.

I think this should be good to go, pending CI results.

stephenmcgruer commented:
Hrm. I don't love that it breaks the assert 'pattern' of having the description be the last argument, but I can't see a better way to do it.

Can you update both the JSDoc in testharness.js and the `docs/` to document this new argument? These methods are getting very complex for developers to use, so good documentation is vital.

bzbarsky commented:
> Hrm. I don't love that it breaks the assert 'pattern' of having the description be the last argument, but I can't see a better way to do it.

We _could_ allow both a `assert_throws_dom(type, func, desc)` and a `assert_throws_dom(type, func, ctor, desc)` form, by examining `typeof` on the third arg in the implementation.  It's not a hard change to make.  Thoughts on that?

> Can you update both the JSDoc in testharness.js and the docs/ to document this new argument? 

Yes, good catch!

stephenmcgruer commented:
> We could allow both a assert_throws_dom(type, func, desc) and a assert_throws_dom(type, func, ctor, desc) form, by examining typeof on the third arg in the implementation. It's not a hard change to make. Thoughts on that?

**I** like it and would say SGTM, but I'm going to coward out and check what @jgraham thinks. I have not spent much time on testharness.js API design.

annevk commented:
Bikeshedding thoughts: we could overload _type_ as well with something like 
```
{ name: "InvalidStateError", ctor: otherGlobal.DOMException }
```
? That'd probably be how we'd this if it were a web platform API. Also, why would we even pass the constructor and not just the global? It always has to be `DOMException` after all. So:
```
{ name: "InvalidStateError", global: otherGlobal }
```

bzbarsky commented:
Yes, we could overload `type`, and if we do that, we could just allow passing `otherGlobal.DOMException("", "InvalidStateError")`, which is less typing than the dictionary form, though arguably a bit less clear in terms of what's really being asserted...

To be very clear, I am quite happy to have this discussion about what the API should look like, because I am not a big fan of what I have so far!

> Also, why would we even pass the constructor and not just the global?

Because of <https://github.com/web-platform-tests/wpt/issues/21809#issuecomment-586600196>: we can't actually pass a _global_, we can only pass a `WindowProxy`.  And that can end up pointing to the wrong global by the time we are doing our comparisons, in tests that perform navigation and then test things in the navigated-away-from global.

bzbarsky commented:
Now we could allow passing _either_ a global or a constructor or something, I guess...  At some point the complexity of the API starts getting in the way of the ergonomics, even if the complexity is arguably in service of the ergonomics.  For example, then people need to understand when they can just pass a global and when that's not enough.

annevk commented:
I'm convinced, no passing of global this. The main oddity with `otherGlobal.DOMException("", "InvalidStateError")` is that in assert_throws_js we pass the constructor and here we'd pass an instance. That's probably okay given that it's only for exceptional cases?

domenic commented:
I'd vote against passing instances.

Here are my preferences:

- Concise-but-perhaps-too-clever: overload `assert_throws_dom(type, func, desc)` and a `assert_throws_dom(type, ctor, func, desc)`. Note keeping the type and ctor adjacent. You would distinguish `ctor` from `func` by whether `arguments[1].name === "DOMException"`, or maybe using [isConstructor](https://esdiscuss.org/topic/isconstructor#content-11).

- Verbose: `assert_throws_dom(type, func, desc)` where `type` is either a string or is `{ name: "InvalidStateError", constructor: otherGlobal.DOMException }`. (maybe `ctor` instead of `constructor`, I dunno.)

I think it's generally OK to be clever in test _utilities_ as long as they are themselves well-tested, and they make the tests easy to write and read. So I lean toward the first.

bzbarsky commented:
I agree passing instances is not great.  It only came up because unlike JS errors the constructor does not uniquely determine the name.

I agree that having cleverness inside the test utilities is fine.

Checking for `isConstructor` does not really work, because if the test function is not an arrow function it's also a constructor.  But checking for `.name === 'DOMException'` would work fine.

I'm personally somewhat in favor of the "four-arg" vs "three-arg" thing, with the first two args being type and ctor in the "four-arg" case (which may only have three args if desc is omitted, of course), so the same thing @domenic likes.  Anyone else want something else here?  If not, I'll update the patch; should be straightforward.

bzbarsky commented:
Alright, that last revision adds docs and implements @domenic's suggestion.

annevk reviewed (approved):
@jgraham should probably merge given the change to testharness.

jgraham reviewed (commented):
This seems OK to me.

I don't love the argument overloading in the API, but if the people writing tests are happy with this design then that seems like the most important thing.

Thanks for adding docs; we unfortunately don't use the jsdoc for anything at the moment so I'm a little worried that the docs in testharness.js itself are better than in the markdown file, but not exposed. BUt maybe the solution there is to improve our docs pipeline to use the jsdoc rather than requiring API docs be duplicated into the markdown files.
